### PR TITLE
Split circuit_input_builder file and improve APIs for it

### PR DESCRIPTION
**Pull request by CPerezz**
As mentioned in https://github.com/appliedzkp/zkevm-circuits/issues/426, CIB file was already more than 3000 lines of code
and was turning into a deposit of all sorts of code for different
purposes.

This is a preliminary work towards adding a bit of order into the
CIB/bus-mapping crate.

After working on this PR, I've realized there's a lot of stuff to improve and tech-debt that we've been leaving unresolved.
So once this PR is merged (to not make an even bigger diff) I'll file issues for each thing I've seen and work on them on separated and easier-to-review PRs.

**NOTE for the reviewers**
The only addition aside from some API getters is the refactor of the `StepAuxiliaryData` and how it works and holds it's variants.
The rest of the PR is simply code splitting and adding some order.

Things to improve in the future:
- CIB Tx struct only used in tests now. Should we migrate away?
- Make all fn's in `error.rs` members of `self` (ie. attribute fns).
- Better handling of `unwrapping` in regular functions.
- Error out instead of panicking in circuit opcodes impl.

### CODE REVIEW

**Comment by ed255** [comment]
The diff for this PR is massive!  Thanks for taking care of this and putting some order :)

I know that the real changes are minimal, but I haven't found a nice tool to analyze the diff easily; so I hope I didn't miss anything.

Here are some thoughts:
1. I see that you introduced getters because now the structs belong to different modules and their private fields need to be accessed from other places.  I'm personally not a big fan of getters when the consumer of the API is ourselves: I think accessing fields looks more clear/transparent than calling methods, and we also avoid the need to have both `foo()` and `foo_mut()`.  So I wanted to propose an idea: what do you think about removing the getters and setting the fields that need to be accessed across modules as `pub(crate)`?  This would affect:
- BlockContext::rwc
- BlockContext::call_map
- ReversionGroup::calls
- ReversionGroup::op_refs
- StepAuxiliaryData::{src_addr, dst_addr, bytes_left, src_addr_end, copy_details}
- CircuitInputBuilderTx::tx_ctx
- TransactionContext::{call_is_success, reversion_groups}
- Transaction::calls
What do you think?

2. I really like the `StepAuxiliaryData` refactor you did!  I think it goes into the right direction, and I think that the `CopyDetails` will be extended in the future; and probably we can find a representation that matches what the future copy circuit would look like.  Here are all the copy operations that I think we will encounter, with the source and destination:
- LOG: memory -> log
- CALLDATACOPY: memory -> memory
- CALLDATACOPY: Tx.CallData -> memory
- (EXT)CODECOPY: bytecode -> memory
- RETURNDATACOPY: memory -> memory
- BeginTx (to = 0), CREATE, CREATE2: memory -> bytecode (this is to always get the code source from the bytecode table)

In summary: overall looks good to me!  Please let me know what you think about point 1, it's OK to disagree :P

**Comment by ed255** [comment]
>     * CIB Tx struct only used in tests now. Should we migrate away?

I don't follow this point.  Do you mean `circuit_input_builder::Transaction`?  If so, this type is exposed in the `CircuitInputStateRef` which the `gen_associated_ops` use.  Maybe you're referring to another type?

>     * Make all fn's in `error.rs` members of `self` (ie. attribute fns).

If you mean `get_step_reported_error`, I think it's a good idea to make it member of `ExecError`.

>     * Better handling of `unwrapping` in regular functions.

This is always good!  I think we should avoid all `unwrapping` in non-test code.  And if it's really needed, at least use `expect` with a small explanation of why it shouldn't fail.

>     * Error out instead of panicking in circuit opcodes impl.

I want to agree on this, but I think we have a challenge which is that halo2 requires it's own error type which makes all errors be `Error::Synthesis` which is very painful to debug.  Nevertheless we could add `log::error` and then return the `Error::Synthesis`, so at least we can trace where the error happened and why.

**Review by ed255 — APPROVED** [review]
LGTM! Thanks for organizing the circuit input builder :)
